internal/daemon/worker/controller_connection_test.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package worker
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	opsservices "github.com/hashicorp/boundary/internal/gen/ops/services"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/nettest"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+func TestMonitorUpstreamConnectionState(t *testing.T) {
+	ctx := context.Background()
+	stateCtx, cancelStateCtx := context.WithCancel(ctx)
+
+	upstreamConnectionState := new(atomic.Value)
+
+	servers, err := createTestServers(t)
+	require.NoError(t, err)
+
+	scheme := strconv.FormatInt(time.Now().UnixNano(), 36)
+	res := manual.NewBuilderWithScheme(scheme)
+	grpcResolver := &grpcResolverReceiver{res}
+
+	cc, err := createTestGRPConnection(res, servers[0].address)
+	require.NoError(t, err)
+
+	// track GRPC state changes
+	go monitorUpstreamConnectionState(stateCtx, cc, upstreamConnectionState)
+
+	grpcResolver.InitialAddresses([]string{servers[0].address})
+
+	t.Cleanup(func() {
+		cc.Close()
+		cancelStateCtx()
+
+		assert.Equal(t, connectivity.Shutdown, cc.GetState())
+
+		for _, s := range servers {
+			s.srv.GracefulStop()
+		}
+	})
+
+	tests := []struct {
+		name             string
+		expectedResponse *opsservices.GetHealthResponse
+		addresses        []string
+		expectedState    connectivity.State
+	}{
+		{
+			name:          "connection with 1 good address",
+			addresses:     []string{servers[0].address},
+			expectedState: connectivity.Ready,
+		},
+		{
+			name:          "connection with multiple good addresses",
+			addresses:     []string{servers[1].address, servers[2].address},
+			expectedState: connectivity.Ready,
+		},
+		{
+			name:          "connection with bad address",
+			addresses:     []string{"bad_address"},
+			expectedState: connectivity.TransientFailure,
+		},
+		{
+			name:          "connection with 1 bad address and 1 good address",
+			addresses:     []string{servers[0].address, "bad_address"},
+			expectedState: connectivity.Ready,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			grpcResolver.SetAddresses(tt.addresses)
+
+			// Add delay for connection state to update with GRPC manual resolver
+			time.Sleep(2 * time.Second)
```

Comment:
A listener as in the server listener or you mean the `monitorUpstreamConnectionState()`? In terms of wait, I realized it takes about 1 second but made it 2 seconds just in case.